internal/dataplane/parser/translators/ingress_atc.go

```diff
@@ -0,0 +1,260 @@
+	if exist && len(snis) > 0 {
+		sniMatcher := sniMatcherFromSNIs(snis)
+		routeMatcher.And(sniMatcher)
+	}
+
+	atc.ApplyExpression(&route.Route, routeMatcher, NormalIngressExpressionPriority)
+	return route
+}
+
+// hostMatcherFromIngressHosts translates hosts in IngressHost format to ATC matcher that matches any of them.
+func hostMatcherFromIngressHosts(hosts []string) atc.Matcher {
+	matchers := make([]atc.Matcher, 0, len(hosts))
+	for _, host := range hosts {
+		if !validHosts.MatchString(host) {
+			continue
+		}
+
+		if strings.HasPrefix(host, "*") {
+			// wildcard match on hosts (like *.foo.com), genreate a suffix match.
+			matchers = append(matchers, atc.NewPrediacteHTTPHost(atc.OpSuffixMatch, strings.TrimPrefix(host, "*")))
+		} else {
+			// exact match on hosts, generate an exact match.
+			matchers = append(matchers, atc.NewPrediacteHTTPHost(atc.OpEqual, host))
+		}
+	}
+	return atc.Or(matchers...)
+}
+
+// pathMatcherFromIngressPath translate ingress path into matcher to match the path.
+func pathMatcherFromIngressPath(httpIngressPath netv1.HTTPIngressPath, regexPathPrefix string) atc.Matcher {
+	switch *httpIngressPath.PathType {
+	// Prefix paths.
+	case netv1.PathTypePrefix:
+		base := strings.Trim(httpIngressPath.Path, "/")
+		if base == "" {
+			// empty string in prefix path matches prefix "/".
+			return atc.NewPredicateHTTPPath(atc.OpPrefixMatch, "/")
+		}
+		return atc.Or(
+			// otherwise, match /<path>/* or /<path>.
+			atc.NewPredicateHTTPPath(atc.OpEqual, "/"+base),
+			atc.NewPredicateHTTPPath(atc.OpPrefixMatch, "/"+base+"/"),
+		)
+	// Exact paths.
+	case netv1.PathTypeExact:
+		relative := strings.TrimLeft(httpIngressPath.Path, "/")
+		return atc.NewPredicateHTTPPath(atc.OpEqual, "/"+relative)
+	// Implementation Specific match. treat it as regex match if it begins with a regex prefix (/~ by default),
+	// otherwise generate a prefix match.
+	case netv1.PathTypeImplementationSpecific:
+		// empty path. matches prefix "/" to match any path.
+		if httpIngressPath.Path == "" {
+			return atc.NewPredicateHTTPPath(atc.OpPrefixMatch, "/")
+		}
+		// regex match.
+		if regexPathPrefix != "" && strings.HasPrefix(httpIngressPath.Path, regexPathPrefix) {
+			regex := strings.TrimPrefix(httpIngressPath.Path, regexPathPrefix)
+			if regex == "" {
+				regex = "^/"
+			}
+			// regex match matches a prefix of the whole path, so we need to add a line start annotation in the regex.
+			if !strings.HasPrefix(regex, "^") {
+				regex = "^" + regex
+			}
+			return atc.NewPredicateHTTPPath(atc.OpRegexMatch, regex)
+		}
+		return atc.NewPredicateHTTPPath(atc.OpPrefixMatch, httpIngressPath.Path)
+	}
+
+	return nil
+}
+
+// protocolMatcherFromProtocols gernerates matchers from protocols.
+func protocolMatcherFromProtocols(protocols []string) atc.Matcher {
+	matchers := []atc.Matcher{}
+	for _, protocol := range protocols {
+		if !util.ValidateProtocol(protocol) {
+			fmt.Println("invalid protocol: ", protocol)
```

Comment:
I assume we don't want to print here but rather handle the error/unwanted state somehow.

If we don't need to handle the error then just going straight to `continue` makes sense, otherwise let's return and `error` and handle it

Reply:
removed the `Println`.